## **CalCOFI Analysis**

**Benedictus Bayu Pramudhito**

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

/usr/local/lib/python3.6/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [2]:
import os
print('\n'.join(os.listdir(r'./drive/My Drive/Colab Test (Bootcamp ML 2020)')))

bottle.csv
References
Untitled0.ipynb
Week_1_Python_1.ipynb
heart.csv
iris.csv
pima-indians-diabetes.csv
sonar.csv
housing.csv
winequality-white.csv
auto.csv


In [3]:
df = pd.read_csv(r'./drive/My Drive/Colab Test (Bootcamp ML 2020)/bottle.csv')
df.head()

/usr/local/lib/python3.6/dist-packages/IPython/core/interactiveshell.py:2718: DtypeWarning: Columns (47,73) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Cst_Cnt  Btl_Cnt       Sta_ID  ... pH2  pH1  DIC Quality Comment
0        1        1  054.0 056.0  ... NaN  NaN                  NaN
1        1        2  054.0 056.0  ... NaN  NaN                  NaN
2        1        3  054.0 056.0  ... NaN  NaN                  NaN
3        1        4  054.0 056.0  ... NaN  NaN                  NaN
4        1        5  054.0 056.0  ... NaN  NaN                  NaN

[5 rows x 74 columns]

In [4]:
df.shape

(864863, 74)

In [5]:
pd.set_option('display.max_rows', 100)
df.dtypes

Cst_Cnt                  int64
Btl_Cnt                  int64
Sta_ID                  object
Depth_ID                object
Depthm                   int64
T_degC                 float64
Salnty                 float64
O2ml_L                 float64
STheta                 float64
O2Sat                  float64
Oxy_µmol/Kg            float64
BtlNum                 float64
RecInd                   int64
T_prec                 float64
T_qual                 float64
S_prec                 float64
S_qual                 float64
P_qual                 float64
O_qual                 float64
SThtaq                 float64
O2Satq                 float64
ChlorA                 float64
Chlqua                 float64
Phaeop                 float64
Phaqua                 float64
PO4uM                  float64
PO4q                   float64
SiO3uM                 float64
SiO3qu                 float64
NO2uM                  float64
NO2q                   float64
NO3uM                  float64
NO3q    

In [6]:
#Numerical data describe
df.describe().T

count           mean  ...          75%          max
Cst_Cnt      864863.0   17138.790958  ...   26557.0000   34404.0000
Btl_Cnt      864863.0  432432.000000  ...  648647.5000  864863.0000
Depthm       864863.0     226.831951  ...     300.0000    5351.0000
T_degC       853900.0      10.799677  ...      13.8800      31.1400
Salnty       817509.0      33.840350  ...      34.1969      37.0340
O2ml_L       696201.0       3.392468  ...       5.5000      11.1300
STheta       812174.0      25.819394  ...      26.6460     250.7840
O2Sat        661274.0      57.103779  ...      97.6000     214.1000
Oxy_µmol/Kg  661268.0     148.808694  ...     240.3796     485.7018
BtlNum       118667.0      10.497426  ...      16.0000      25.0000
RecInd       864863.0       4.700273  ...       7.0000       7.0000
T_prec       853900.0       2.017148  ...       2.0000       3.0000
T_qual        23127.0       7.471829  ...       9.0000       9.0000
S_prec       817509.0       2.716787  ...       3.0000       3.0000
S_qual        74914.0       7.945057  ...       9.0000       9.0000
P_qual       673755.0       8.999996  ...       9.0000       9.0000
O_qual       184676.0       8.755615  ...       9.0000       9.0000
SThtaq        65823.0       8.467010  ...       9.0000       9.0000
O2Satq       217797.0       8.793487  ...       9.0000       9.0000
ChlorA       225272.0       0.450185  ...       0.3900      66.1100
Chlqua       639166.0       8.999848  ...       9.0000       9.0000
Phaeop       225271.0       0.198553  ...       0.2300      65.3000
Phaqua       639170.0       8.999844  ...       9.0000       9.0000
PO4uM        413317.0       1.564747  ...       2.4800       5.2100
PO4q         451786.0       8.999320  ...       9.0000       9.0000
SiO3uM       354091.0      26.606587  ...      41.5400     196.0000
SiO3qu       510866.0       8.999687  ...       9.0000       9.0000
NO2uM        337576.0       0.042316  ...       0.0300       8.1900
NO2q         529474.0       8.995458  ...       9.0000       9.0000
NO3uM        337403.0      17.302247  ...      30.0000      95.0000
NO3q         529933.0       8.993122  ...       9.0000       9.0000
NH3uM         64962.0       0.084877  ...       0.0600      15.6300
NH3q         808299.0       8.947668  ...       9.0000       9.0000
C14As1        14432.0       9.757704  ...       8.0000     584.5000
C14A1p        12760.0       1.266928  ...       2.0000       2.0000
C14A1q       848605.0       8.999992  ...       9.0000       9.0000
C14As2        14414.0       9.762613  ...       8.0565     948.3000
C14A2p        12742.0       1.267776  ...       2.0000       2.0000
C14A2q       848623.0       8.999992  ...       9.0000       9.0000
DarkAs        22649.0       0.155295  ...       0.1740       6.9000
DarkAp        20457.0       1.993450  ...       2.0000       2.0000
DarkAq       840440.0       8.999992  ...       9.0000       9.0000
MeanAs        22650.0       8.393530  ...       7.0000     948.3000
MeanAp        20457.0       1.256587  ...       2.0000       2.0000
MeanAq       840439.0       8.999992  ...       9.0000       9.0000
LightP        18651.0      18.356509  ...      24.0000      99.9000
R_Depth      864863.0     226.832495  ...     300.0000    5351.0000
R_TEMP       853900.0      10.799697  ...      13.8800      31.1400
R_POTEMP     818816.0      10.842295  ...      13.9200      31.1400
R_SALINITY   817509.0      33.840295  ...      34.1960      37.0340
R_SIGMA      812007.0      25.811010  ...      26.6400     250.7840
R_SVA        812092.0     220.938918  ...     299.8000     683.4000
R_DYNHT      818206.0       0.431763  ...       0.6400       3.8800
R_O2         696201.0       3.392498  ...       5.5000      11.1300
R_O2Sat      666448.0      57.192940  ...      97.6000     214.1000
R_SIO3       354099.0      26.606862  ...      41.5000     196.0000
R_PO4        413325.0       1.564736  ...       2.4800       5.2100
R_NO3        337411.0      17.302188  ...      30.0000      95.0000
R_NO

In [7]:
#Categorical Data
df.describe(include='O').T

count  ...   freq
Sta_ID               864863  ...  10043
Depth_ID             864863  ...      2
IncTim                14437  ...    325
DIC Quality Comment      55  ...      9

[4 rows x 4 columns]

In [8]:
#Cek Missing Values
nulls =  df.isnull().sum().sort_values(ascending=False)
percnt = round(nulls/len(df)*100, 2)

df_null = pd.DataFrame(columns=['Features', 'Total', 'Percent'])
df_null['Features'] = nulls.index
df_null['Total'] = nulls.values
df_null['Percent'] = percnt.values

#Cek attribut dengan missing values > dari 30%
df_null

Features   Total  Percent
0                   pH2  864853   100.00
1   DIC Quality Comment  864808    99.99
2                   pH1  864779    99.99
3                  DIC2  864639    99.97
4                   TA2  864629    99.97
5                  DIC1  862864    99.77
6                   TA1  862779    99.76
7                C14A2p  852121    98.53
8                C14A1p  852103    98.52
9                C14As2  850449    98.33
10               C14As1  850431    98.33
11               IncTim  850426    98.33
12               LightP  846212    97.84
13               DarkAp  844406    97.63
14               MeanAp  844406    97.63
15               DarkAs  842214    97.38
16               MeanAs  842213    97.38
17               T_qual  841736    97.33
18                NH3uM  799901    92.49
19                R_NH4  799881    92.49
20               SThtaq  799040    92.39
21               S_qual  789949    91.34
22               BtlNum  746196    86.28
23               R_SAMP  742857    85.89
24               O_qual  680187    78.65
25               O2Satq  647066    74.82
26               Phaeop  639592    73.95
27               ChlorA  639591    73.95
28              R_PHAEO  639588    73.95
29               R_CHLA  639587    73.95
30                NO3uM  527460    60.99
31                R_NO3  527452    60.99
32                NO2uM  527287    60.97
33                R_NO2  527279    60.97
34               SiO3uM  510772    59.06
35               R_SIO3  510764    59.06
36                PO4uM  451546    52.21
37                R_PO4  451538    52.21
38                 PO4q  413077    47.76
39               SiO3qu  353997    40.93
40                 NO2q  335389    38.78
41                 NO3q  334930    38.73
42               Chlqua  225697    26.10
43               Phaqua  225693    26.10
44          Oxy_µmol/Kg  203595    23.54
45                O2Sat  203589    23.54
46              R_O2Sat  198415    22.94
47               P_qual  191108    22.10
48               O2ml_L  168662    19.50
49                 R_O2  168662    19.50
50                 NH3q   56564     6.54
51              R_SIGMA   52856     6.11
52                R_SVA   52771     6.10
53               STheta   52689     6.09
54               Salnty   47354     5.48
55               S_prec   47354     5.48
56           R_SALINITY   47354     5.48
57              R_DYNHT   46657     5.39
58             R_POTEMP   46047     5.32
59               MeanAq   24424     2.82
60               DarkAq   24423     2.82
61               C14A1q   16258     1.88
62               C14A2q   16240     1.88
63               R_TEMP   10963     1.27
64               T_prec   10963     1.27
65               T_degC   10963     1.27
66               RecInd       0     0.00
67              R_Depth       0     0.00
68               R_PRES       0     0.00
69               Depthm       0     0.00
70             Depth_ID       0     0.00
71               Sta_ID       0     0.00
72              Btl_Cnt       0     0.00
73              Cst_Cnt       0     0.00

In [9]:
for i in df_null['Features'][df_null['Percent'] > 30]:
  df = df.drop([i], axis=1)
  print(df.shape, i)

(864863, 73) pH2
(864863, 72) DIC Quality Comment
(864863, 71) pH1
(864863, 70) DIC2
(864863, 69) TA2
(864863, 68) DIC1
(864863, 67) TA1
(864863, 66) C14A2p
(864863, 65) C14A1p
(864863, 64) C14As2
(864863, 63) C14As1
(864863, 62) IncTim
(864863, 61) LightP
(864863, 60) DarkAp
(864863, 59) MeanAp
(864863, 58) DarkAs
(864863, 57) MeanAs
(864863, 56) T_qual
(864863, 55) NH3uM
(864863, 54) R_NH4
(864863, 53) SThtaq
(864863, 52) S_qual
(864863, 51) BtlNum
(864863, 50) R_SAMP
(864863, 49) O_qual
(864863, 48) O2Satq
(864863, 47) Phaeop
(864863, 46) ChlorA
(864863, 45) R_PHAEO
(864863, 44) R_CHLA
(864863, 43) NO3uM
(864863, 42) R_NO3
(864863, 41) NO2uM
(864863, 40) R_NO2
(864863, 39) SiO3uM
(864863, 38) R_SIO3
(864863, 37) PO4uM
(864863, 36) R_PO4
(864863, 35) PO4q
(864863, 34) SiO3qu
(864863, 33) NO2q
(864863, 32) NO3q


In [10]:
#Isi Missing Values lainnya dengan mode values
for i in df.columns:
  if df[i].isnull().sum() > 0:
    df[i].fillna(df[i].mode().mean(), inplace=True)
    print('filled', i)

filled T_degC
filled Salnty
filled O2ml_L
filled STheta
filled O2Sat
filled Oxy_µmol/Kg
filled T_prec
filled S_prec
filled P_qual
filled Chlqua
filled Phaqua
filled NH3q
filled C14A1q
filled C14A2q
filled DarkAq
filled MeanAq
filled R_TEMP
filled R_POTEMP
filled R_SALINITY
filled R_SIGMA
filled R_SVA
filled R_DYNHT
filled R_O2
filled R_O2Sat


In [11]:
#Cek ulang Missing Valuesnya
nulls = df.isnull().sum().sort_values(ascending=False)
percnt = round(nulls/len(df)*100, 2)

df_null = pd.DataFrame(columns=['Features', 'Total', 'Percent'])
df_null['Features'] = nulls.index
df_null['Total'] = nulls.values
df_null['Percent'] = percnt.values

df_null

Features  Total  Percent
0        R_PRES      0      0.0
1       R_O2Sat      0      0.0
2       Btl_Cnt      0      0.0
3        Sta_ID      0      0.0
4      Depth_ID      0      0.0
5        Depthm      0      0.0
6        T_degC      0      0.0
7        Salnty      0      0.0
8        O2ml_L      0      0.0
9        STheta      0      0.0
10        O2Sat      0      0.0
11  Oxy_µmol/Kg      0      0.0
12       RecInd      0      0.0
13       T_prec      0      0.0
14       S_prec      0      0.0
15       P_qual      0      0.0
16       Chlqua      0      0.0
17       Phaqua      0      0.0
18         NH3q      0      0.0
19       C14A1q      0      0.0
20       C14A2q      0      0.0
21       DarkAq      0      0.0
22       MeanAq      0      0.0
23      R_Depth      0      0.0
24       R_TEMP      0      0.0
25     R_POTEMP      0      0.0
26   R_SALINITY      0      0.0
27      R_SIGMA      0      0.0
28        R_SVA      0      0.0
29      R_DYNHT      0      0.0
30         R_O2      0      0.0
31      Cst_Cnt      0      0.0

In [12]:
#Cek Nilai Duplikatnya
print('Terdapat {} Nilai Duplikat pada Dataset'.format(df.duplicated().sum()))

Terdapat 0 Nilai Duplikat pada Dataset


In [13]:
df['Depth_ID'].head()

0    19-4903CR-HY-060-0930-05400560-0000A-3
1    19-4903CR-HY-060-0930-05400560-0008A-3
2    19-4903CR-HY-060-0930-05400560-0010A-7
3    19-4903CR-HY-060-0930-05400560-0019A-3
4    19-4903CR-HY-060-0930-05400560-0020A-7
Name: Depth_ID, dtype: object

In [14]:
#Extract Month/Year dari Depth_ID
#Depth_ID = [Century]-[YY][MM][ShipCode]-etc
#19-4903CR-HY-060-0930-05400560-0000A-3

df['Year'] = (df['Depth_ID'].str.split('-', expand=True)[0] + df['Depth_ID'].str.split('-', expand=True)[1]).map(lambda x: str(x)[:4])
df['Month'] = (df['Depth_ID'].str.split('-', expand=True)[1]).map(lambda x: str(x)[2:4])

df[['Depth_ID', 'Year', 'Month']].head()

Depth_ID  Year Month
0  19-4903CR-HY-060-0930-05400560-0000A-3  1949    03
1  19-4903CR-HY-060-0930-05400560-0008A-3  1949    03
2  19-4903CR-HY-060-0930-05400560-0010A-7  1949    03
3  19-4903CR-HY-060-0930-05400560-0019A-3  1949    03
4  19-4903CR-HY-060-0930-05400560-0020A-7  1949    03

In [15]:
df_1 = df.copy()
print(df.shape)
df_1.head(2)

(864863, 34)


Cst_Cnt  Btl_Cnt       Sta_ID  ... R_PRES  Year  Month
0        1        1  054.0 056.0  ...      0  1949     03
1        1        2  054.0 056.0  ...      8  1949     03

[2 rows x 34 columns]

In [16]:
drop_cols = ['Cst_Cnt', 'Btl_Cnt', 'Sta_ID', 'Depth_ID', 'Year', 'Month']
df = df.drop(drop_cols, axis=1)
df.shape

(864863, 28)

**Machine Learning Model**

**Data Preprocessing**

1. Lakukan preprocessing terhadap data menggunakan:
  - StandardScaler
  - Yeo-Johnson (PowerTransformer)
  - Box-Cox Transformation (PowerTransformer)
  - QuantileTransformer
  - MinMaxScaler
  - MaxAbsScaler
  - RobustScaler

  Dan jelaskan apabila hasil training untuk algoritma LinearRegression untuk masing-masing metode transformasi data dibandingkan!

  Jelaskan juga apabila terdapat metode preprocessing yang tidak dapat digunakan untuk salah satu dataset!

In [17]:
df.describe()

Depthm         T_degC  ...        R_O2Sat         R_PRES
count  864863.000000  864863.000000  ...  864863.000000  864863.000000
mean      226.831951      10.778132  ...      67.564246     228.395694
std       316.050259       4.221126  ...      37.678563     319.456731
min         0.000000       1.440000  ...      -0.100000       0.000000
25%        46.000000       7.720000  ...      31.400000      46.000000
50%       125.000000       9.990000  ...      83.900000     126.000000
75%       300.000000      13.830000  ...     102.400000     302.000000
max      5351.000000      31.140000  ...     214.100000    5458.000000

[8 rows x 28 columns]

In [18]:
from sklearn.preprocessing import StandardScaler, PowerTransformer, QuantileTransformer, MinMaxScaler, MaxAbsScaler, RobustScaler

In [19]:
stdr = StandardScaler()
powbox = PowerTransformer(method='yeo-johnson', standardize=True)
powbox1 = PowerTransformer(method='box-cox', standardize=True)
quan = QuantileTransformer()
minmax = MinMaxScaler()
maxabs = MaxAbsScaler()
rbst = RobustScaler()

In [20]:
X = df.drop(['T_degC'], axis=1)
y = df['T_degC']

In [21]:
X_stdr = stdr.fit_transform(X.values)
y_stdr = stdr.fit_transform(df['T_degC'].values.reshape(-1,1)).flatten()

X_powbox = powbox.fit_transform(X.values)
y_powbox = powbox.fit_transform(df['T_degC'].values.reshape(-1,1)).flatten()

#X_powbox1 = powbox1.fit_transform(X.values)
#y_powbox1 = powbox1.fit_transform(df['T_degC'].values.reshape(-1, 1)).flatten()

X_quan = quan.fit_transform(X.values)
y_quan = quan.fit_transform(df['T_degC'].values.reshape(-1,1)).flatten()

X_minmax = minmax.fit_transform(X.values)
y_minmax = minmax.fit_transform(df['T_degC'].values.reshape(-1,1)).flatten()

X_maxabs = maxabs.fit_transform(X.values)
y_maxabs = maxabs.fit_transform(df['T_degC'].values.reshape(-1,1)).flatten()

X_rbst = rbst.fit_transform(X.values)
y_rbst = rbst.fit_transform(df['T_degC'].values.reshape(-1,1)).flatten()

/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:195: RuntimeWarning: overflow encountered in multiply
  x = um.multiply(x, x, out=x)
/usr/local/lib/python3.6/dist-packages/numpy/core/_methods.py:199: RuntimeWarning: overflow encountered in reduce
  ret = umr_sum(x, axis, dtype, out, keepdims)


Power Transformers dengan metode Box-Cox tidak dapat digunakan karena pada dataset ditemukan data dengan nilai <= 0

### **Linear Regression**

**Standard Scaler**

In [22]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X_stdr, y_stdr, test_size=0.25, random_state=10)
X_train.shape, X_test.shape, y_train.shape, y_test.shape

((648647, 27), (216216, 27), (648647,), (216216,))

In [23]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score
from sklearn.linear_model import LinearRegression

In [24]:
linreg = LinearRegression()
linreg.fit(X_train, y_train)
prediksi_linreg = linreg.predict(X_test)

In [25]:
mse = mean_squared_error(y_test, prediksi_linreg)
mae = mean_absolute_error(y_test, prediksi_linreg)
r2 = r2_score(y_test, prediksi_linreg)
print('MSE (Mean Squared Error) Using Standard Scaler is  :', mse)
print('MAE (Mean Absolute Error) Using Standard Scaler is :', mae)
print('r^2 score Using Standard Scaler is                 :', r2)
print('RMSE (Root Mean Squared Error) is                  :', np.sqrt(mean_squared_error(y_test, prediksi_linreg)))

MSE (Mean Squared Error) Using Standard Scaler is  : 4.727345152370335e-08
MAE (Mean Absolute Error) Using Standard Scaler is : 2.7624532670652526e-05
r^2 score Using Standard Scaler is                 : 0.9999999528980487
RMSE (Root Mean Squared Error) is                  : 0.0002174245881304673


**Power Transformer (yeo-johnson)**

In [26]:
X_train_yeo, X_test_yeo, y_train_yeo, y_test_yeo = train_test_split(X_powbox, y_powbox, test_size=0.25, random_state=10)
X_train_yeo.shape, X_test_yeo.shape, y_train_yeo.shape, y_test_yeo.shape

((648647, 27), (216216, 27), (648647,), (216216,))

In [27]:
linreg_yeo = LinearRegression()
linreg_yeo.fit(X_train_yeo, y_train_yeo)
prediksi_linreg_yeo = linreg_yeo.predict(X_test_yeo)

In [28]:
mse = mean_squared_error(y_test_yeo, prediksi_linreg_yeo)
mae = mean_absolute_error(y_test_yeo, prediksi_linreg_yeo)
r2 = r2_score(y_test_yeo, prediksi_linreg_yeo)
print('MSE (Mean Squared Error) Using Power Transformer is        :', mse)
print('MAE (Mean Absolute Error) Using Power Transformer is       :', mae)
print('r^2 score Using Power Transformer is                       :', r2)
print('RMSE (Root Mean Squared Error) Using Power Transformer is  :', np.sqrt(mean_squared_error(y_test_yeo, prediksi_linreg_yeo)))

MSE (Mean Squared Error) Using Power Transformer is        : 1.9458842838290586e-08
MAE (Mean Absolute Error) Using Power Transformer is       : 2.795653812175824e-05
r^2 score Using Power Transformer is                       : 0.9999999805776073
RMSE (Root Mean Squared Error) Using Power Transformer is  : 0.00013949495631846546


**Quantile Transformer**

In [29]:
X_train_quan, X_test_quan, y_train_quan, y_test_quan = train_test_split(X_quan, y_quan, test_size=0.25, random_state=10)
X_train_quan.shape, X_test_quan.shape, y_train_quan.shape, y_test_quan.shape

((648647, 27), (216216, 27), (648647,), (216216,))

In [30]:
linreg_quan = LinearRegression()
linreg_quan.fit(X_train_quan, y_train_quan)
prediksi_linreg_quan = linreg_quan.predict(X_test_quan)

In [31]:
mse = mean_squared_error(y_test_quan, prediksi_linreg_quan)
mae = mean_absolute_error(y_test_quan, prediksi_linreg_quan)
r2 = r2_score(y_test_quan, prediksi_linreg_quan)
print('MSE (Mean Squared Error) Using Quantile Transformer is        :', mse)
print('MAE (Mean Absolute Error) Using Quantile Transformer is       :', mae)
print('r^2 score Using Quantile Transformer is                       :', r2)
print('RMSE (Root Mean Squared Error) Using Quantile Transformer is  :', np.sqrt(mean_squared_error(y_test_quan, prediksi_linreg_quan)))

MSE (Mean Squared Error) Using Quantile Transformer is        : 1.2212740306554458e-06
MAE (Mean Absolute Error) Using Quantile Transformer is       : 0.0008458099523296782
r^2 score Using Quantile Transformer is                       : 0.9999853164700966
RMSE (Root Mean Squared Error) Using Quantile Transformer is  : 0.0011051126778095733


**MinMax Scaler**

In [32]:
X_train_minmax, X_test_minmax, y_train_minmax, y_test_minmax = train_test_split(X_minmax, y_minmax, test_size=0.25, random_state=10)

In [33]:
linreg_minmax = LinearRegression()
linreg_minmax.fit(X_train_minmax, y_train_minmax)
prediksi_linreg_minmax = linreg_minmax.predict(X_test_minmax)

In [34]:
mse = mean_squared_error(y_test_minmax, prediksi_linreg_minmax)
mae = mean_absolute_error(y_test_minmax, prediksi_linreg_minmax)
r2 = r2_score(y_test_minmax, prediksi_linreg_minmax)
print('MSE (Mean Squared Error) Using MinMax Scaler is        :', mse)
print('MAE (Mean Absolute Error) Using MinMax Scaler is       :', mae)
print('r^2 score Using MinMax Scaler is                       :', r2)
print('RMSE (Root Mean Squared Error) Using MinMax Scaler is  :', np.sqrt(mean_squared_error(y_test_minmax, prediksi_linreg_minmax)))

MSE (Mean Squared Error) Using MinMax Scaler is        : 9.54905987007975e-10
MAE (Mean Absolute Error) Using MinMax Scaler is       : 3.926147329504752e-06
r^2 score Using MinMax Scaler is                       : 0.9999999528980487
RMSE (Root Mean Squared Error) Using MinMax Scaler is  : 3.09015531487978e-05


**MaxAbs Scaler**

In [35]:
X_train_maxabs, X_test_maxabs, y_train_maxabs, y_test_maxabs = train_test_split(X_maxabs, y_maxabs, test_size=0.25, random_state=10)
X_train_maxabs.shape, X_test_maxabs.shape, y_train_maxabs.shape, y_test_maxabs.shape

((648647, 27), (216216, 27), (648647,), (216216,))

In [36]:
linreg_maxabs = LinearRegression()
linreg_maxabs.fit(X_train_maxabs, y_train_maxabs)
prediksi_linreg_maxabs = linreg_maxabs.predict(X_test_maxabs)

In [37]:
mse = mean_squared_error(y_test_maxabs, prediksi_linreg_maxabs)
mae = mean_absolute_error(y_test_maxabs, prediksi_linreg_maxabs)
r2 = r2_score(y_test_maxabs, prediksi_linreg_maxabs)
print('MSE (Mean Squared Error) Using MaxAbs Scaler is        :', mse)
print('MAE (Mean Absolute Error) Using MaxAbs Scaler is       :', mae)
print('r^2 score Using MaxAbs Scaler is                       :', r2)
print('RMSE (Root Mean Squared Error) Using MaxAbs Scaler is  :', np.sqrt(mean_squared_error(y_test_maxabs, prediksi_linreg_maxabs)))

MSE (Mean Squared Error) Using MaxAbs Scaler is        : 8.686329478528703e-10
MAE (Mean Absolute Error) Using MaxAbs Scaler is       : 3.7445913734861593e-06
r^2 score Using MaxAbs Scaler is                       : 0.9999999528980487
RMSE (Root Mean Squared Error) Using MaxAbs Scaler is  : 2.9472579592782004e-05


**Robust Scaler**

In [38]:
X_train_rbst, X_test_rbst, y_train_rbst, y_test_rbst = train_test_split(X_rbst, y_rbst, test_size=0.25, random_state=10)
X_train_rbst.shape, X_test_rbst.shape, y_train_rbst.shape, y_test_rbst.shape

((648647, 27), (216216, 27), (648647,), (216216,))

In [39]:
linreg_rbst = LinearRegression()
linreg_rbst.fit(X_train_rbst, y_train_rbst)
prediksi_linreg_rbst = linreg_rbst.predict(X_test_rbst)

In [40]:
mse = mean_squared_error(y_test_rbst, prediksi_linreg_rbst)
mae = mean_absolute_error(y_test_rbst, prediksi_linreg_rbst)
r2 = r2_score(y_test_rbst, prediksi_linreg_rbst)
print('MSE (Mean Squared Error) Using Robust Scaler is        :', mse)
print('MAE (Mean Absolute Error) Using Robust Scaler is       :', mae)
print('r^2 score Using Robust Scaler is                       :', r2)
print('RMSE (Root Mean Squared Error) Using Robust Scaler is  :', np.sqrt(mean_squared_error(y_test_rbst, prediksi_linreg_rbst)))

MSE (Mean Squared Error) Using Robust Scaler is        : 2.2562701323524054e-08
MAE (Mean Absolute Error) Using Robust Scaler is       : 1.9084545908298203e-05
r^2 score Using Robust Scaler is                       : 0.9999999528980487
RMSE (Root Mean Squared Error) Using Robust Scaler is  : 0.00015020885900480057


Keenam preprocessor mampu memberikan nilai r2 yang sangat tinggi dan juga nilai rmse yang mendekati nol saat diuji terhadap testing data dengan memakai model linear regression.

Nilai r2 diperoleh sebesar 0.99999 dengan keenam preprocessor. Sementara nilai rmse nya bervariasi untuk keenam preprocessor namun keenamnya memperoleh nilai rmse yang sangat mendekati nol 

2. Gunakan algoritma berikut dan lakukan percobaan hyperparameter tuning untuk mendapatkan hasil terbaik:
  - Random Forest Regressor
  - XGBoost
  - SVM Regressor
  - KNN Regressor

  Hitung angka r^2 serta RMSE untuk menentukan mana algoritma yang paling baik!

**XGBoost Regressor**

In [41]:
from xgboost import XGBRegressor

In [47]:
xgb = XGBRegressor(objective='reg:squarederror', colsample_bytree=0.5, learning_rate=0.03, max_depth=4, alpha=10, n_estimators=200)
xgb.fit(X_train_maxabs, y_train_maxabs)
prediksi_xgb = xgb.predict(X_train_maxabs)
prediksi_xgb_test = xgb.predict(X_test_maxabs)
r2 = r2_score(y_test_maxabs, prediksi_xgb_test)
print('r^2 score Using MaxAbs Scaler is :', r2)
print('Nilai RMSE XGBoost Regressor pada Testing Set adalah : ', np.sqrt(mean_squared_error(prediksi_xgb_test, y_test_maxabs)))
print('================================================================================================================================================================)')

r^2 score Using MaxAbs Scaler is : 0.9991362521071713
Nilai RMSE XGBoost Regressor pada Testing Set adalah :  0.003991096236502418
================================================================================================================================================================)


**Support Vector Regressor**

In [44]:
from sklearn.svm import SVR

In [45]:
svr = SVR()
svr.get_params()

{'C': 1.0,
 'cache_size': 200,
 'coef0': 0.0,
 'degree': 3,
 'epsilon': 0.1,
 'gamma': 'scale',
 'kernel': 'rbf',
 'max_iter': -1,
 'shrinking': True,
 'tol': 0.001,
 'verbose': False}

In [46]:
svr = SVR(epsilon=0.05)
svr.fit(X_train_maxabs, y_train_maxabs)
prediksi_svr_test = svr.predict(X_test_maxabs)
r2 = r2_score(y_test_maxabs, prediksi_svr_test)
print('r^2 score Using MaxAbs Scaler is :', r2)
print('Nilai RMSE SVC dengan Kernel rbf adalah : ', np.sqrt(mean_squared_error(y_test_maxabs, prediksi_svr_test)))

r^2 score Using MaxAbs Scaler is : 0.9655243280730921
Nilai RMSE SVC dengan Kernel rbf adalah :  0.025214773374594233


**Random Forest Regressor**

In [48]:
from sklearn.ensemble import RandomForestRegressor

In [49]:
rf = RandomForestRegressor()
rf.get_params()

{'bootstrap': True,
 'ccp_alpha': 0.0,
 'criterion': 'mse',
 'max_depth': None,
 'max_features': 'auto',
 'max_leaf_nodes': None,
 'max_samples': None,
 'min_impurity_decrease': 0.0,
 'min_impurity_split': None,
 'min_samples_leaf': 1,
 'min_samples_split': 2,
 'min_weight_fraction_leaf': 0.0,
 'n_estimators': 100,
 'n_jobs': None,
 'oob_score': False,
 'random_state': None,
 'verbose': 0,
 'warm_start': False}

In [50]:
rf = RandomForestRegressor(n_estimators=50, random_state=10)
rf.fit(X_train_maxabs, y_train_maxabs)
prediksi_rf_test = rf.predict(X_test_maxabs)
r2 = r2_score(y_test_maxabs, prediksi_rf_test)
print('r^2 score Using MaxAbs Scaler is :', r2)
print('RMSE untuk model Random Forest dengan n_estimator 50 adalah :', np.sqrt(mean_squared_error(y_test_maxabs, prediksi_rf_test)))

r^2 score Using MaxAbs Scaler is : 0.9999996687854658
RMSE untuk model Random Forest dengan n_estimator 1600 adalah : 7.815439266721137e-05


**K-Nearest Neighbors Regressor**

In [ ]:
from sklearn.neighbors import KNeighborsRegressor

In [ ]:
neig = np.arange(1, 10, 2)
scorer2 = []

for i, k in enumerate(neig):
  knn = KNeighborsRegressor(n_neighbors=k)
  knn.fit(X_train_maxabs, y_train_maxabs)
  prediksi_knn_test = knn.predict(X_test_maxabs)
  scorer2.append(r2_score(y_test_maxabs, prediksi_knn_test))

plt.figure(figsize=[10,8])
plt.plot(neig, scorer2, label = 'Testing R2')
plt.legend()
plt.title('-value VS R2')
plt.xlabel('Number of Neighbors')
plt.ylabel('R2')
plt.xticks(neig)
plt.grid()
plt.show()
print("\nNilai R2 terbaik untuk Model KNN adalah {}".format(round(np.max(scorer2), 2)))

Dengan model Random Forest, Support Vector Regressor, dan XGBoost Regressor diperoleh nilai rmse yang sangat mendekati nol dan nilai r2 yang sangat tinggi (yakni diatas 0.95), dengan nilai r2 tertinggi diperoleh dengan model Linear Regression, XGBoost Regressor, dan Random Forest Regressor. 

Untuk model KNN belum sempat dirun jadi tidak diketahui nilai r2 dan rmsenya